cmd/controller/controller.go

```diff
@@ -283,30 +281,92 @@ func getReleaseName(r *helmCrdV1.HelmRelease) string {
 	return rname
 }
 
+func findIndex(target string, s []string) int {
+	res := -1
+	for i := range s {
+		if s[i] == target {
+			res = i
+		}
+	}
+	return res
+}
+func pull(item string, s []string) ([]string, error) {
+	index := findIndex(item, s)
+	if index == -1 {
+		return []string{}, fmt.Errorf("%s not present in %v", item, s)
+	}
+	return append(s[:index], s[index+1:]...), nil
+}
+func hasFinalizer(h *helmCrdV1.HelmRelease) bool {
+	currentFinalizers := h.ObjectMeta.Finalizers
+	for _, f := range currentFinalizers {
+		if f == releaseFinalizer {
+			return true
+		}
+	}
+	return false
+}
+
+func removeFinalizer(helmReleaseClient helmClientset.Interface, helmObj *helmCrdV1.HelmRelease) error {
+	helmObjClone := helmObj.DeepCopy()
+	newSlice, _ := pull(releaseFinalizer, helmObj.ObjectMeta.Finalizers)
+	if len(newSlice) == 0 {
+		newSlice = nil
+	}
+	helmObjClone.ObjectMeta.Finalizers = newSlice
+	_, err := helmReleaseClient.HelmV1().HelmReleases(helmObj.Namespace).Update(helmObjClone)
```

Comment:
I think it will be cleaner to separate updating the in-memory HelmRelease object, and _actually_ doing the `Update()` API call to update the server-side object.  In particular, that easily allows testing, and "batching" several changes into a single `Update()`.